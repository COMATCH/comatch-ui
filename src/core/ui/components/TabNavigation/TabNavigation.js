import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import noop from 'lodash/noop';

import './TabNavigation.scss';

/**
 * A tab navigation that can appear above a `Panel`.
 *
 * When the tabs are clicked they will usually reveal different parts of the content.
 */

export class TabNavigation extends Component {
    static propTypes = {
        // eslint-disable-next-line valid-jsdoc
        /**
         * Determines which tab should be active
         */
        activeTabIndex: (props, propName, componentName) => {
            const prop = props[propName];
            if (typeof prop !== 'number') {
                return new Error(
                    `Invalid prop \`${propName}\` of type \`${typeof prop}\` supplied to` +
                        ` \`${componentName}\`. Expected number.`,
                );
            }
            if (Array.isArray(props.children) && prop > props.children.length) {
                return new Error(
                    `Invalid prop \`${propName}\` supplied to` +
                        ` \`${componentName}\`. Value of \`${propName}\` ` +
                        `must be less or equal to props.children.length`,
                );
            }
            return false;
        },
        children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
        /**
         * Updates the state with the active tab before invoking `onTabClick`
         */
        handleTabClick: PropTypes.func,
        onTabClick: PropTypes.func,
    };

    static defaultProps = {
        activeTabIndex: 0,
        children: null,
        handleTabClick: null,
        onTabClick: noop,
    };

    state = {
        activeTabIndex: this.props.activeTabIndex,
    };

    componentDidUpdate(prevProps) {
        if (this.props.activeTabIndex !== prevProps.activeTabIndex) {
            this.setState({ activeTabIndex: this.props.activeTabIndex });
        }
    }

    handleTabClick = (index) => {
        this.setState({ activeTabIndex: index });
        this.props.onTabClick();
    };

    render() {
        const { children } = this.props;
        const { activeTabIndex } = this.state;
        const handleTabClick = this.props.handleTabClick || this.handleTabClick;

        return (
            <nav className="TabNavigation">
                <ul className="TabNavigation__list">
                    {React.Children.map(children, (child, key) => {
                        const classes = classNames('TabNavigation__list-item', {
                            active: activeTabIndex === key,
                        });

                        return (
                            <li className={classes} key={key} onClick={() => handleTabClick(key)}>
                                {child}
                            </li>
                        );
                    })}
                </ul>
            </nav>
        );
    }
}
