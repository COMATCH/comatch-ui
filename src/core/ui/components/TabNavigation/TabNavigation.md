
## TabNavigation

From [`TabNavigation.js`](TabNavigation.js)

A tab navigation that can appear above a `Panel`.

When the tabs are clicked they will usually reveal different parts of the content.

prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**activeTabIndex** | `(custom validator)` | `0` | :x: | Determines which tab should be active
**children** | `Union<Array[]<ReactNode> \| ReactNode>` | `null` | :x: | 
**handleTabClick** | `Function` | `null` | :x: | Updates the state with the active tab before invoking `onTabClick`
**onTabClick** | `Function` | `noop` | :x: | 



