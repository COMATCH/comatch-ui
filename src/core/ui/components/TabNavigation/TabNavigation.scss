@import '../../../styles/variables.scss';

.TabNavigation {
    button {
        outline: 0;
        background: none;
        border: none;
    }

    &__list {
        margin: 0;
        padding: 0;
    }

    // Elements
    &__list-item {
        background-color: $co-color-white;
        border: 1px solid $co-color-mid-gray;
        border-top-left-radius: 3px;
        border-top-right-radius: 3px;
        // display
        display: inline-block;
        // positioning
        position: relative;
        top: 1px;
        z-index: 1;

        // Links contained in tabs
        button,
        a:link,
        a:active,
        a:hover,
        a:visited {
            color: $co-color-black;
            cursor: pointer;
            display: inline-block;
            padding: 8px 10px;
            transition: color 250ms ease-out;
            // text
            font-size: $co-font-md;
            font-weight: 500;
            text-align: center;
            text-decoration: none;
            text-transform: uppercase;
        }

        // States
        &.active {
            border-bottom-color: $co-color-white;

            button,
            a:link,
            a:active,
            a:hover,
            a:visited {
                color: $co-color-primary;
            }
        }
    }

    // Whitespace arrangement
    &__list-item {
        &:first-child button,
        &:first-child a {
            // Extra whitespace for first item to align with
            // panel contents.
            padding-left: 20px;
        }

        & ~ .TabNavigation__list-item {
            // Whitespace between subsequent tabs
            margin-left: 5px;
        }
    }

    // Reset the top margin of Panels following a TabNavigation
    & ~ .Panel {
        border-top-left-radius: 0;
        margin-top: 0;
    }
}
