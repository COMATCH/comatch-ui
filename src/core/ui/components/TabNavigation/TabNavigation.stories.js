/* eslint-disable import/no-unresolved */
import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { TabNavigation } from './TabNavigation';
import { Panel } from '../Panel/Panel';

const panelText = `Lorem ipsum dolor sit amet, consectetur adipiscing elit.
Pellentesque porttitor efficitur enim, vitae blandit risus hendrerit non.
Donec in dignissim felis. Aenean eget risus non est suscipit finibus.
Fusce sed ipsum dapibus mi feugiat dapibus vel vel libero. Phasellus varius.`;

const onTabClick = () => {
    action('clicked tab')();
};

const onLinkClick = (evt) => {
    evt.preventDefault();
    onTabClick();
};

storiesOf('TabNavigation', module)
    .add('with default props, rendering buttons', () => (
        <div>
            <TabNavigation>
                <button onClick={onTabClick}>Tab 1</button>
                <button onClick={onTabClick}>Tab 2</button>
                <button onClick={onTabClick}>Tab 3</button>
            </TabNavigation>
            <Panel title="A panel title" titleColor="green">
                {panelText}
            </Panel>
        </div>
    ))
    .add('with default props, rendering links', () => (
        <div>
            <TabNavigation>
                <a href="#" onClick={onLinkClick}>
                    Tab 1
                </a>
                <a href="#" onClick={onLinkClick}>
                    Tab 2
                </a>
                <a href="#" onClick={onLinkClick}>
                    Tab 3
                </a>
            </TabNavigation>
            <Panel title="A panel title" titleColor="green">
                {panelText}
            </Panel>
        </div>
    ))
    .add('rendering links and activeTabIndex set to 2', () => (
        <div>
            <TabNavigation activeTabIndex={2}>
                <a href="#" onClick={onLinkClick}>
                    Tab 1
                </a>
                <a href="#" onClick={onLinkClick}>
                    Tab 2
                </a>
                <a href="#" onClick={onLinkClick}>
                    Tab 3
                </a>
            </TabNavigation>
            <Panel title="A panel title" titleColor="green">
                {panelText}
            </Panel>
        </div>
    ));
