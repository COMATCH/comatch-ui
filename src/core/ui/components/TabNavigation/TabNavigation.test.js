import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import { TabNavigation } from './TabNavigation';
import '../../../config/tests/setup';

describe('TabNavigation', () => {

    // if no activeTabIndex passed the first tab should be active (covered by the snapshot)
    // should render each child in separate <li> element (covered by the snapshot)

    it('should render correctly', () => {
        const tabNavigation = renderer.create(
            <TabNavigation>
                <button>Tab 1</button>
                <button>Tab 2</button>
                <button>Tab 3</button>
            </TabNavigation>
        ).toJSON();
        expect(tabNavigation).toMatchSnapshot();
    });

    it('should render correctly with default activeTabIndex', () => {
        const tabNavigation = renderer.create(
            <TabNavigation activeTabIndex={2}>
                <button>Tab 1</button>
                <button>Tab 2</button>
                <button>Tab 3</button>
            </TabNavigation>
        ).toJSON();
        expect(tabNavigation).toMatchSnapshot();
    });

    it('should change the active tab on click', () => {
        const wrapper = mount(
            <TabNavigation>
                <button>Tab 1</button>
                <button>Tab 2</button>
                <button>Tab 3</button>
            </TabNavigation>
        );

        let secondElem = wrapper.find('.TabNavigation__list-item').at(2);
        secondElem.simulate('click');

        wrapper.update();
        secondElem = wrapper.find('.TabNavigation__list-item').at(2);

        expect(secondElem.prop('className')).toEqual('TabNavigation__list-item active');
    });
});
