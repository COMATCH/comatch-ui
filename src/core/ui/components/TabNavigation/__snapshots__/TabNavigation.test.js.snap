// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabNavigation should render correctly 1`] = `
<nav
  className="TabNavigation"
>
  <ul
    className="TabNavigation__list"
  >
    <li
      className="TabNavigation__list-item active"
      onClick={[Function]}
    >
      <button>
        Tab 1
      </button>
    </li>
    <li
      className="TabNavigation__list-item"
      onClick={[Function]}
    >
      <button>
        Tab 2
      </button>
    </li>
    <li
      className="TabNavigation__list-item"
      onClick={[Function]}
    >
      <button>
        Tab 3
      </button>
    </li>
  </ul>
</nav>
`;

exports[`TabNavigation should render correctly with default activeTabIndex 1`] = `
<nav
  className="TabNavigation"
>
  <ul
    className="TabNavigation__list"
  >
    <li
      className="TabNavigation__list-item"
      onClick={[Function]}
    >
      <button>
        Tab 1
      </button>
    </li>
    <li
      className="TabNavigation__list-item"
      onClick={[Function]}
    >
      <button>
        Tab 2
      </button>
    </li>
    <li
      className="TabNavigation__list-item active"
      onClick={[Function]}
    >
      <button>
        Tab 3
      </button>
    </li>
  </ul>
</nav>
`;
